### PR TITLE
Implement atomic write for tweets

### PR DESCRIPTION
**Pull request by timofurrer**
Implement atomic writing for the local tweets.

Refs #34

### CODE REVIEW

File: twtxt/helper.py [hunk 1]
```diff
@@ -85,3 +87,28 @@ def sort_and_truncate_tweets(tweets, direction, limit):
...
+def atomic_write(path, mode):
+    """
+        Simple contextmanager to safely write to a
+        file.
+
+        :param str filename: the path to write to
+        :param str mode: the mode to open the file
+    """
+    # generate random filepath in same location as given filename
+    tmpfilepath = os.path.join(os.path.dirname(path), ".{}".format(uuid.uuid4()))
```
Comment (kitchen) [inline twtxt/helper.py]: Is there a way to leverage python's built-in tempfile support for this? https://docs.python.org/2/library/tempfile.html

It's possible there may not be. The temporary file has to be on the same filesystem as the 'real' file, and I don't know if tempfile has bits that can ensure that or not. Just a suggestion!

File: twtxt/helper.py [hunk 2]
```diff
@@ -85,3 +87,28 @@ def sort_and_truncate_tweets(tweets, direction, limit):
...
+    try:
+        if "a" in mode:
+            os.rename(path, tmpfilepath)
+
+        with open(tmpfilepath, mode) as atomicfile:
+            yield atomicfile
+        os.rename(tmpfilepath, path)
```
Comment (kitchen) [inline twtxt/helper.py]: I'm reading this as "rename the file to tempfile, write to it, rename the file back". I don't think this does what you want?

File: twtxt/helper.py [hunk 3]
```diff
@@ -85,3 +87,28 @@ def sort_and_truncate_tweets(tweets, direction, limit):
...
+    finally:
+        try:
+            os.unlink(tmpfilepath)
```
Comment (kitchen) [inline twtxt/helper.py]: This, combined with above could lead to dropping your tweetfile on the floor if there's a write failure.

File: twtxt/helper.py (hunk 1 above)
Comment (timofurrer) [inline twtxt/helper.py]: That was exactly the problem. For the rename it must be on the same partition - on my system the /tmp is on a different one - that's why I preferred the way with uuid. In addition the tempfile comes really handy with the contextmanagers and self-deletion however, I would have to disabled those.

File: twtxt/helper.py (hunk 2 above)
Comment (timofurrer) [inline twtxt/helper.py]: I don't see your point. 
First I'm copying (if append mode) the existing tweets to a tempfile. After that I open it and yield the file object. The tweet can be written. After that the file will be closed and I'll copy the the new content back to the original location. Did I miss something?

File: twtxt/helper.py (hunk 3 above)
Comment (timofurrer) [inline twtxt/helper.py]: Oh yeah, I see. The first os.rename would need to be *rolled back* .. somehow ..

File: twtxt/helper.py (hunk 2 above)
Comment (kitchen) [inline twtxt/helper.py]: My point is you aren't copying the file, just renaming it. Unless os.rename does something different in Python than I would expect :D

**Comment by buckket** [comment]
Click as a custom open_file() method which allows atomic behaviour: http://click.pocoo.org/6/api/#click.open_file.

Maybe we can make use of it.

**Comment by timofurrer** [comment]
I'll have a look at clicks `open_file`.

**Comment by timofurrer** [comment]
click does not support atomic appending. See https://github.com/mitsuhiko/click/blob/master/click/_compat.py#L434
They suggest to open in w-mode and copy data explicitly.. What do you think about that?

**Comment by kitchen** [comment]
Yea, that's basically what you have to do. Copy the file, append to new file, rename back.

-Jeremy

--
Insert PGP signature here.

**Comment by timofurrer** [comment]
I've updated the PR. You can easily test it with this:

```
>>> with twtxt.helper.atomic_write("testfile", "a") as f:
...     f.write("Hallo")
...     raise AttributeError("FAILED")
... 
5
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
AttributeError: FAILED
>>> 
```

**Comment by timofurrer** [comment]
@kitchen do you agree with this implementation?

**Pull request closed without merging**